Cache gcc warnings

facchinm opened the pull request:
... and display them even if the file is not being recompiled.
Should allow removing 48aa042 from #301  
There's a concurrency issue on cached prints that should be solved before merging.

matthijskooijman reviewed (changes requested):
Overall, this looks good to me. I did leave quite some smaller remarks inline.

I don't understand the concurrency issue exactly: If this interleaves output from multiple concurrent processes, then why doesn't that happen in the original compilation run? Is that perhaps just a coincidence, that the chances of interleaving are small when each compilation command actually takes some time? Or is there some mechanism in place to actually properly interleave output from multiple processes? I can actually imagine that, for multi-process compilation, it would make sense to (only) capture all output from a single compilation command and then write that output atomically (rather than letting the compiler write to os.stdout/stderr directly, with the risk of interleaving)?

facchinm commented on utils/utils.go:
```diff
@@ -291,33 +292,62 @@ func ExecCommand(ctx *types.Context, command *exec.Cmd, stdout int, stderr int)
+		go func() {
+			for stdOut.Scan() {
+				txt := stdOut.Text() + "\n"
+				fmt.Fprint(os.Stderr, txt)
```
Yes, my fault, copy paste stuff

facchinm commented on utils/utils.go:
```diff
@@ -291,33 +292,62 @@ func ExecCommand(ctx *types.Context, command *exec.Cmd, stdout int, stderr int)
 		command.Stdout = buffer
 	} else if stdout == Show || stdout == ShowIfVerbose && ctx.Verbose {
-		command.Stdout = os.Stdout
+		stdoutCopy, _ = command.StdoutPipe()
```
Totally agree, I only tried to modify the needed pieces but a complete refactor wouldn't hurt.

facchinm commented on utils/utils.go:
```diff
@@ -291,33 +292,62 @@ func ExecCommand(ctx *types.Context, command *exec.Cmd, stdout int, stderr int)
+		stdOut := bufio.NewScanner(stdoutCopy)
+		stdErr := bufio.NewScanner(stderrCopy)
+		stdOut.Split(bufio.ScanLines)
+		stdErr.Split(bufio.ScanLines)
```
need to take a closer look at bufio but there are probably better implementations anyway

facchinm commented on builder_utils/utils.go:
```diff
@@ -230,12 +230,18 @@ func compileFileWithRecipe(ctx *types.Context, sourcePath string, source string,
 
 	if !objIsUpToDate {
-		_, _, err = ExecRecipe(ctx, properties, recipe, false /* stdout */, utils.ShowIfVerbose /* stderr */, utils.Show)
+		_, stderrCache, err := ExecRecipe(ctx, properties, recipe, false /* stdout */, utils.ShowIfVerbose /* stderr */, utils.Show)
```
:+1:

facchinm commented:
@matthijskooijman there's a mutex that prevents interleaving when printing via `logger` stream, but the fact that warnings are not getting interleaved right now looks like pure luck (or something I can't understand) :slightly_smiling_face: .
However, compilers are quite normal beasts, they don't mess up with terminal magic like some uploaders do, so I believe we can safely cache everything and print atomically when the command returns.

matthijskooijman commented on builder_utils/utils.go:
```diff
@@ -230,12 +230,18 @@ func compileFileWithRecipe(ctx *types.Context, sourcePath string, source string,
+			logger.Println(constants.LOG_LEVEL_INFO, constants.MSG_USING_PREVIOUS_COMPILED_FILE, properties[constants.BUILD_PROPERTIES_OBJECT_FILE])
+		}
+		if ctx.WarningsCache[source] != "" {
+			logger.UnformattedFprintln(os.Stderr, ctx.WarningsCache[source])
```
We can still have a single map, just put a dict / struct inside with Stderr and Stdout members? It really seems a pity to only implement half of this, and then later realize that there's a compiler that does generate stdout output...

To test, you could make the compiler generate stdout output by adding -fdump-tree-original=stdout to the commandline.

facchinm replied:
Done in 4006605

facchinm commented:
@matthijskooijman I think I addressed all you concerns. At the end the interleaving only happens when the streams are crunched by the Java IDE (I'll take a closer look later), but they're fine when launched on a terminal.
Thanks a lot for the review!

matthijskooijman reviewed (commented):
Looks good. Two more minor remarks added.

Are you planning to squash some of thes commits together before merging?

facchinm committed:
Refactor ExecCommand to cache stdout/stderr

This commits doesn't change the functionality of ExecCommand EXCEPT for programs that show progress bars or similar.
For this kind of utilities, please use Stream "type"

facchinm committed:
Introduce OutputCache file

This file should cache warnings and other compiler output between compilations.
Directly maps to ctx.OutputCache map

facchinm committed:
Display cached warnings for previously compiled files

The mutex is needed because the map could be concurrently accessed (when using parallel compilation)
